src/java/htsjdk/tribble/index/IndexFactory.java

```diff
@@ -165,16 +165,19 @@ public static IndexType getIndexType(final BufferedInputStream is) {
      * @param indexFile from which to load the index
      */
     public static Index loadIndex(final String indexFile) {
```

Comment:
This still seems ugly.  I propose the following instead:

```java
   /**
     * Load in index from the specified file.   The type of index (LinearIndex or IntervalTreeIndex) is determined
     * at run time by reading the type flag in the file.
     *
     * @param indexFile from which to load the index
     */
    public static Index loadIndex(final String indexFile) {
        try (InputStream inputStream = openIndexFile(indexFile);
             // Must be buffered, because getIndexType uses mark and reset
             BufferedInputStream bufferedInputStream = new BufferedInputStream(inputStream, Defaults.NON_ZERO_BUFFER_SIZE))
        {
            final Class<Index> indexClass = IndexType.getIndexType(bufferedInputStream).getIndexType();

            final Constructor<Index> ctor = indexClass.getConstructor(InputStream.class);

            return ctor.newInstance(bufferedInputStream);
        } catch (final IOException ex) {
            throw new TribbleException.UnableToReadIndexFile("Unable to read index file", indexFile, ex);
        } catch (final Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    private static InputStream openIndexFile(final String indexFile) throws IOException {
        final InputStream inputStreamInitial = ParsingUtils.openInputStream(indexFile);
        if (indexFile.endsWith(".gz")) {
            return new GZIPInputStream(inputStreamInitial);
        }
        else if (indexFile.endsWith(TabixUtils.STANDARD_INDEX_EXTENSION)) {
            return new BlockCompressedInputStream(inputStreamInitial);
        }
        else {
            return inputStreamInitial;
        }
    }
```

also needs `Class getIndexType()` ->` Class<Index> getIndexType()`